### ISSUE TITLE
Selecting a better metric for the "best" model

### ISSUE THREAD

**Issue by Tom-Forsyth** [ultralytics/ultralytics#19188]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi!

I am using YOLO11 segmentation with the large model. For my use case, I have 3-4 large objects in each image, and reasoning which class each belongs to is very obvious, so I never have issues with precision or recall. I also don't care about bounding boxes. The only thing I care about is how accurate are the segmentation masks.

Often times, the best model (from a mAP standpoint) does not produce the best segmentation masks. I normally will train a certain amount of epochs, save the model at each epoch, and select the weights from the epoch with the lowest validation segmentation loss, and deploy that model.

This works, but is there a way to be able to change my "best" evaluation metric? Is there an even better metric for my application than validation seg loss?

Thanks!

**Comment by glenn-jocher** [comment]
👋 Hello @Tom-Forsyth, thank you for your interest in Ultralytics 🚀! Your question is insightful, and we appreciate you sharing the details of your use case.

We recommend checking out the [Ultralytics Documentation](https://docs.ultralytics.com/) for guidance on evaluating models and other insights that may help you determine the best metric for your application's needs. 

If this is a 🐛 bug or specific feature request, providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) would help us better understand the context and assist further. For experimentation related to metrics or custom evaluation methods, sharing any additional logs or experiment details would be very helpful.

### Custom Metric Guidance
While currently YOLO models save the "best" checkpoint by default based on mAP, you can modify your training loop or validation process to track and evaluate against other metrics, like segmentation loss or a custom-designed metric. Advanced modifications might require changes to the training code. Feel free to share more details of your experimentation if you'd like further support.

### Community Support
For further insights or to discuss new ideas, join the Ultralytics community where it suits you best:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- Forums for detailed discussions: [Discourse](https://community.ultralytics.com/)
- Casual knowledge sharing: [Subreddit](https://reddit.com/r/Ultralytics)

### Version and Environments
Make sure you're using the latest version of the `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) and dependencies installed. You can upgrade with the following:

```bash
pip install -U ultralytics
```

For executing your experiments, you might find these verified environments helpful (📓 Notebooks with GPUs included):
- <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> 
- <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> 
- <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

If you're interested in cloud or containerized solutions, see our guides for [Google Cloud](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/), [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/), and [Docker](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

### CI Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing to ensure compatibility across all supported platforms.

🚀 This is an automated response, but an Ultralytics engineer will follow up with you soon to assist further. Thank you for your patience!

**Comment by Y-T-G** [comment]
You can change the fitness function to prioritize segmentation.

https://github.com/ultralytics/ultralytics/blob/9d8c2fe3c7a2701a9cdc769b4b8786738a5b99d0/ultralytics/utils/metrics.py#L1007

But the box is important for segmentation because it's used to crop the mask.

**Comment by Tom-Forsyth** [comment]
> You can change the fitness function to prioritize segmentation.
> 
> [ultralytics/ultralytics/utils/metrics.py](https://github.com/ultralytics/ultralytics/blob/9d8c2fe3c7a2701a9cdc769b4b8786738a5b99d0/ultralytics/utils/metrics.py#L1007)
> 
> Line 1007 in [9d8c2fe](/ultralytics/ultralytics/commit/9d8c2fe3c7a2701a9cdc769b4b8786738a5b99d0)
>  return self.seg.fitness() + self.box.fitness() 
> 
> But the box is important for segmentation because it's used to crop the mask.

Great, thanks! Is there any way to provide a metric as a whole to consider the best? For example, if I want to use validation loss or a custom metric.

**Comment by Y-T-G** [comment]
You have access to mAP, recall and precision in that particular file. Flr validation loss, you would need to modify the trainer.

https://github.com/ultralytics/ultralytics/blob/b9b315a1e2af006c23da4cd88cf7f16015ea7432/ultralytics/engine/trainer.py#L433

**Comment by Tom-Forsyth** [comment]
Thanks!

**Comment by glenn-jocher** [comment]
For custom segmentation-focused metrics, you can modify the validation loss calculation in the training loop by overriding the `on_fit_epoch_end` callback to track your preferred metric. The segmentation loss components are calculated in `v8SegmentationLoss` (see [loss.py](https://docs.ultralytics.com/reference/utils/loss/#ultralytics.utils.loss.v8SegmentationLoss)), which you could adapt to prioritize mask accuracy. You might consider implementing a custom IoU calculation specifically for your masks. Would you like guidance on implementing a specific modification?
